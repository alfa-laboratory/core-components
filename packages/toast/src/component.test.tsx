import React, { forwardRef } from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as popoverModule from '@alfalab/core-components-popover';
import { Toast, ToastProps } from './index';

describe('Toast', () => {
    jest.useFakeTimers();

    const baseProps = {
        open: true,
        onClose: jest.fn(),
    };

    const getPortalContainer = () => {
        if (document.querySelector('#portalContainer')) {
            return document.querySelector('#portalContainer') as HTMLElement;
        }

        const portalContainer = document.createElement('div');
        portalContainer.setAttribute('id', 'portalContainer');
        document.body.appendChild(portalContainer);
        return portalContainer;
    };

    describe('Snapshots tests', () => {
        it('should match snapshot when open', () => {
            const { baseElement } = render(<Toast {...baseProps}>text</Toast>);

            expect(baseElement).toMatchSnapshot();
        });

        it('should match snapshot when closed', () => {
            const { baseElement } = render(
                <Toast {...baseProps} open={false}>
                    text
                </Toast>,
            );

            expect(baseElement).toMatchSnapshot();
        });

        it('should math snapshot when prop `anchorElement` is passed', () => {
            const anchorElement = document.createElement('div');
            document.body.appendChild(anchorElement);

            const { baseElement } = render(
                <Toast {...baseProps} anchorElement={anchorElement}>
                    text
                </Toast>,
            );

            expect(baseElement).toMatchSnapshot();
        });
    });

    it('should set `className` class', () => {
        const className = 'test-class';

        render(
            <Toast {...baseProps} className={className} getPortalContainer={getPortalContainer} />,
        );

        expect(getPortalContainer().firstElementChild).toHaveClass(className);
    });

    it('should pass props to Popover', () => {
        const popoverComponentSpy = jest.spyOn(popoverModule, 'Popover');

        const anchorElement = document.createElement('div');
        document.body.appendChild(anchorElement);

        const popoverProps: Partial<popoverModule.PopoverProps> = {
            position: 'top',
            offset: [5, 5],
            open: true,
            getPortalContainer,
            anchorElement,
            /*
             * todo: почему-то preventFlip и transition не прокидываются, хотя указаны в типах пропсов. Нужно либо прокинуть либо убрать из типов
             * preventFlip: false,
             * transition: { timeout: 200 }
             */
        };

        render(<Toast {...baseProps} anchorElement={anchorElement} {...popoverProps} />);

        const popoverLastCall =
            popoverComponentSpy.mock.calls[popoverComponentSpy.mock.calls.length - 1];

        expect(popoverLastCall[0]).toMatchObject(popoverProps);
    });

    it('should set bottomOffset', () => {
        render(<Toast {...baseProps} bottomOffset={10} getPortalContainer={getPortalContainer} />);

        const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

        expect(toastPlate).toHaveStyle('bottom: 10px');
    });

    it('should forward ref to ToastPlate wrapper', () => {
        const ref = jest.fn();

        render(<Toast {...baseProps} ref={ref} />);

        expect(ref.mock.calls[0][0].tagName).toBe('DIV');
    });

    describe('Callback tests', () => {
        it('should call onClose when click outside', () => {
            const onClose = jest.fn();
            render(<Toast onClose={onClose} open={true} getPortalContainer={getPortalContainer} />);

            userEvent.click(getPortalContainer().firstElementChild as HTMLElement);
            expect(onClose).not.toBeCalled();

            userEvent.click(document.body);
            expect(onClose).toBeCalled();
        });

        it('should call onClose after delay', () => {
            const onClose = jest.fn();
            render(<Toast onClose={onClose} open={true} autoCloseDelay={3000} />);

            jest.advanceTimersByTime(2500);
            expect(onClose).not.toBeCalled();

            jest.advanceTimersByTime(3500);
            expect(onClose).toBeCalled();
        });

        it('should not call onClose if mouse is over ToastPlate', () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            userEvent.hover(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).not.toBeCalled();
        });

        it('should call onClose after mouse left ToastPlate', () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            userEvent.hover(toastPlate);
            userEvent.unhover(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).toBeCalled();
        });

        it('should not call onClose if touch ToastPlate', () => {
            const onClose = jest.fn();
            render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                />,
            );

            const toastPlate = getPortalContainer().firstElementChild as HTMLElement;

            jest.advanceTimersByTime(2500);
            fireEvent.touchStart(toastPlate);
            jest.advanceTimersByTime(5000);

            expect(onClose).not.toBeCalled();
        });

        it('should render custom toast plate', async () => {
            const dataTestId = 'testId';
            const onClose = jest.fn();

            const CustomToastPlate: ToastProps['ToastPlate'] = forwardRef((props, ref) => (
                <div {...props} ref={ref} data-test-id={dataTestId} />
            ));

            const { getByTestId } = render(
                <Toast
                    onClose={onClose}
                    open={true}
                    autoCloseDelay={3000}
                    getPortalContainer={getPortalContainer}
                    ToastPlate={CustomToastPlate}
                />,
            );

            await waitFor(() => expect(getByTestId(dataTestId)).toBeInTheDocument());
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Toast {...baseProps} />);

        expect(unmount).not.toThrowError();
    });
});
