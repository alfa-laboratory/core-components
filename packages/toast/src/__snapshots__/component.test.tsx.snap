// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Snapshots tests should match snapshot when closed 1`] = `
<body>
  <div
    alfa-portal-container=""
  />
  <div />
</body>
`;

exports[`Toast Snapshots tests should match snapshot when open 1`] = `
<body>
  <div />
  <div
    alfa-portal-container=""
  >
    <div
      class="component fixed toastPlate"
      style="z-index: 1000;"
    >
      <div
        class="contentWrap"
      >
        <div
          class="content"
        >
          <div>
            <div
              class="children"
            >
              text
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Toast Snapshots tests should math snapshot when prop \`anchorElement\` is passed 1`] = `
<body>
  <div
    alfa-portal-container=""
  >
    <div
      class="component"
      style="z-index: 1000; position: absolute; left: 0px; top: 0px;"
    >
      <div
        class="inner popoverInner"
        style="transition-duration: 150ms;"
      >
        <div
          class=""
        >
          <div
            class="component"
          >
            <div
              class="contentWrap"
            >
              <div
                class="content"
              >
                <div>
                  <div
                    class="children"
                  >
                    text
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
  <div />
</body>
`;
